inductor stride assertion: make error message more actionable

bdhirsh opened the pull request:
This assertion has tripped enough times that we know of a few culprits. Some of these culprits are toggleable via config, so I updated the error message to mention that they're worth turning off if you hit the assertion:

(1) layout optimization

(2) ddp optimizer

(3) (technically) this issue, although hopefully it will be fixed soon: https://github.com/pytorch/pytorch/issues/116433

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #116874
* #117587



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

bdhirsh committed:
inductor stride assertion: make error message more actionable

[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge

bdhirsh committed:
Update on "inductor stride assertion: make error message more actionable"

This assertion has tripped enough times that we know of a few culprits. Some of these culprits are toggleable via config, so I updated the error message to mention that they're worth turning off if you hit the assertion:

(1) layout optimization

(2) ddp optimizer

(3) (technically) this issue, although hopefully it will be fixed soon: https://github.com/pytorch/pytorch/issues/116433




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

albanD commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -478,7 +478,18 @@ def codegen_input_size_asserts(self):
             stride = self.codegen_shape_tuple(buf.get_stride())
-            self.prefix.writeline(f"assert_size_stride({name}, {size}, {stride})")
+            error_msg = """'''
+Found a size/stride mismatch. Try re-running with layout optimization disabled:
```
```suggestion
Found a size/stride mismatch between eager mode and the Tensor generated by the inductor backend. Try re-running with layout optimization disabled:
```

albanD reviewed (approved):
Small nit, but sounds good

bdhirsh commented:
TODO: also mention optimize_ddp_lazy_compile in the error message

wconstab commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -478,7 +478,18 @@ def codegen_input_size_asserts(self):
+    `TORCHINDUCTOR_LAYOUT_OPTIMIZATION=0` (env var)
+If you are using DDP, and the assert is still failing even with layout optimization disabled,
+then please additionally try disabling the DDP Optimizer:
+    `torch._dynamo.config.optimize_ddp = False`
```
best to suggest (1) turn on the lazy ddp comple flag, then if that fails, (2) optimize_ddp=False

bdhirsh replied:
Will do!

eellison commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -478,7 +478,18 @@ def codegen_input_size_asserts(self):
-            self.prefix.writeline(f"assert_size_stride({name}, {size}, {stride})")
+            error_msg = """'''
+Found a size/stride mismatch. Try re-running with layout optimization disabled:
+You can do this with either:
```
We should be able to detect if we are in a ddp optimizer setting, right? Can we avoid the long error message if this is not the case.

bdhirsh committed:
Update on "inductor stride assertion: make error message more actionable"

This assertion has tripped enough times that we know of a few culprits. Some of these culprits are toggleable via config, so I updated the error message to mention that they're worth turning off if you hit the assertion:

(1) layout optimization

(2) ddp optimizer

(3) (technically) this issue, although hopefully it will be fixed soon: https://github.com/pytorch/pytorch/issues/116433




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

bdhirsh committed:
Update on "inductor stride assertion: make error message more actionable"

This assertion has tripped enough times that we know of a few culprits. Some of these culprits are toggleable via config, so I updated the error message to mention that they're worth turning off if you hit the assertion:

(1) layout optimization

(2) ddp optimizer

(3) (technically) this issue, although hopefully it will be fixed soon: https://github.com/pytorch/pytorch/issues/116433




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

bdhirsh committed:
Update on "inductor stride assertion: make error message more actionable"

This assertion has tripped enough times that we know of a few culprits. Some of these culprits are toggleable via config, so I updated the error message to mention that they're worth turning off if you hit the assertion:

(1) layout optimization

(2) ddp optimizer

(3) (technically) this issue, although hopefully it will be fixed soon: https://github.com/pytorch/pytorch/issues/116433




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

wconstab commented:
are we dropping this @bdhirsh or is it still relevant?
